Added heavy transactions and optimized some code

abhi3700 opened the pull request:
## Description:

This PR introduces code for sending heavy transactions to the Nova network. Also, some code optimization has been done in addition to logging improvements when running the CLI app.

## Key Changes:

- <u>Logging Improvements</u>: Added logging via `log::info!` instead of printing via `println!` to console to avoid cluttering the console.

  <details>
  <summary><b>View CLI output for looking at all info:</b></summary>

  ```sh
  $ RUST_LOG=info cargo run -- \
  --funding-amount 441000000000000 \
  --initial-funded-account-private-key $FUNDER_PRIVATE_KEY \
  --num-accounts 3 \
  --rpc-url $SUBSPACE_EVM_RPC_URL \
  --transaction-type light

  Funder's initial balance: 195.999968145065149880 TSSC.
  =====

  Address[0]:     0x29b9fd654d763770515263d6aed389c4bdcaff16
  Private key[0]: 0xcedab660bf7c993cb1ebb1edce6803829b0e46a841ee34b79de33971364e9eb0

  Address[1]:     0x907472cff03f041123c8d2074352b7c75290619b
  Private key[1]: 0x31b96888f6f730108c3a4b07699c4027431ab1fafbd647e61450cdd3c3c88a90

  Address[2]:     0xb1ff3cc7675a322b89837693261be391bb337482
  Private key[2]: 0xd19abf298c2a77e8c8a4625af0a5688bee759d5dd7dfca4db764ebb507ea6109

  Calling 'Fund' contract's 'transferTsscToMany' method for sending funds in bulk...

  '0x0370…d246' sent funds to newly created accounts, which incurred a gas fee of '0.0004092300027282 TSSC', has a tx hash: '0xec3eab729e01a600a4704a9b6a2bcf2a88912621da2ffaf8a4d2f6a59db01d7a', indexed at #0 in block #63645.

  Sending light transactions...
  [2023-11-05T13:54:40Z INFO  dtp::utils] Number stored in 'Counter' before calls: 13341

  [2023-11-05T13:55:10Z INFO  dtp::utils] '0x9074…619b' increment number, which incurred a gas fee of '0.00007928100052854 TSSC', has a tx hash: '0x2959baea82e42afb28f1f27afe76e0cf3840d5174b3ca3a8cd1037781f7cdbb5', indexed at #0 in block #63648.

  [2023-11-05T13:55:10Z INFO  dtp::utils] '0x29b9…ff16' increment number, which incurred a gas fee of '0.00007928100052854 TSSC', has a tx hash: '0x520a1abd73096630fc993e5a72b5dfef406da140e6fd20495737e2169d1447c6', indexed at #1 in block #63648.

  [2023-11-05T13:55:10Z INFO  dtp::utils] '0xb1ff…7482' increment number, which incurred a gas fee of '0.00007928100052854 TSSC', has a tx hash: '0x3f42fefa762d0ac82fa9d8d5f1349b4322e34b84c758f29c7e35e95061a71656', indexed at #2 in block #63648.

  [2023-11-05T13:55:11Z INFO  dtp::utils] Number stored in 'Counter' after 3 calls: 13344

  Light transactions sent successfully.
  =====
  Funder's final balance: 195.998245359062484640 TSSC.
  Funder spent: 0.001722786002665240 TSSC
  ```

  </details>
  <details>
  <summary><b>View output as CLI app:</b></summary>

  ```sh
  $ cargo run -- \
  --funding-amount 441000000000000 \
  --initial-funded-account-private-key $FUNDER_PRIVATE_KEY \
  --num-accounts 3 \
  --rpc-url $SUBSPACE_EVM_RPC_URL \
  --transaction-type light

  Funder's initial balance: 195.998245359062484640 TSSC.
  =====

  Address[0]:     0x3ad845e47eff630f49a0e70f1f92a7048591f1d7
  Private key[0]: 0x24c8ef7c8327785826f3c1f2a5cbec239c8c03496c5a166551e4f5c2289ada2f

  Address[1]:     0x786f349fff9832aa041e1efcb5503ccdea02e15c
  Private key[1]: 0xe9f6b33aeff72f443c42675d0d9cd4c7d75e3337aaa4ec061c25f526dfef540a

  Address[2]:     0x37f0526f9dc235bd2fec4a1fc0a29d7b968f0cb4
  Private key[2]: 0x7af1781adf5f3bfe88aa40538b4121569babbd8a6437f7444aba64f362101a64

  Calling 'Fund' contract's 'transferTsscToMany' method for sending funds in bulk...

  '0x0370…d246' sent funds to newly created accounts, which incurred a gas fee of '0.0004092300027282 TSSC', has a tx hash: '0x695d1d264246e1d726a979b4fc4e80043c0074b57cc08bfa4f77bfaf2ada0dc1', indexed at #0 in block #63662.

  Sending light transactions...
  Light transactions sent successfully.
  =====
  Funder's final balance: 195.996522573059819400 TSSC.
  Funder spent: 0.001722786002665240 TSSC
  ```

  </details>

- <u>Code optimization</u>:
  - Add chain id to `.env.example` to avoid async call to get the value of chain id in iterations.
  - Archived all old unused code, cleaned up the codebase removing all unused imports, variables, functions, etc.
  - Memory preallocation: Replacing `Vec::new()` with `Vec::with_capacity()` in all the places to pre-allocate the exact amount of space needed to prevent reallocation as the vector grows.
  - Cloned client arc rather than owning. `client.to_owned()` modified to `client.clone()` as it was giving 'value moved here' error.
  - Removed cloning client Arc when calling `get_gas_cost` function, just reference it.
  - Contracts functions have been separated into a separate module for better code readability
- Added code for sending heavy transactions to the Nova network.

abhi3700 committed:
Added batch handling for async calls in Load
contract's setArray method

abhi3700 committed:
Optimized code by calling chain id directly from .env

- Add logging and debugging libraries
- Add chain id to '.env.example' to avoid async call to get the value of chain id in iterations.
- Archived all old unused code

abhi3700 committed:
Code optimization

- cleaned up the codebase removing all unused imports, variables, functions, etc.
- Memory preallocation: Replacing 'Vec::new()' with 'Vec::with_capacity()' in all the places to pre-allocate the exact amount of space needed to prevent reallocation as the vector grows.
- Cloned client arc rather than owning. 'client.to_owned()' modified to 'client.clone()' as it was giving 'value moved here' error.
- Removed cloning client Arc when calling 'get_gas_cost' function, just reference it.

abhi3700 committed:
Contracts functions have been separated into a separate module for better code readability

abhi3700 committed:
Added constants as configurable via environment variables

- MAX_BATCH_SIZE
- MAX_LOAD_COUNT_PER_BLOCK

abhi3700 committed:
Added code for sending heavy txs in main.rs, logs/

abhi3700 committed:
Fixed adding an arg for function to send heavy tx, logs/

abhi3700 committed:
updated submodule w a generic test function to test for any value of count

0xjeremyfrank commented on dtp/src/main.rs:
```diff
@@ -77,8 +76,19 @@ async fn main() -> Result<()> {
+                load_address,
+                multicall_address,
+                fund_contract_addr,
+                chain_id,
```
You could just set `chain_id` from a single call to the rpc rather than force the user to add it as an environment variable. The chain id is going to have to match what the rpc provides anyway. The big improvement from prior implementation is to not make the call on every signed transaction.

abhi3700 replied:
Fair enough. I will do the required changes.

abhi3700 replied:
[commit](https://github.com/abhi3700/domain-transaction-producer/pull/5/commits/45f07ad16d8f96946129a30f1a0da2e076e6c0e6)

0xjeremyfrank commented on dtp/src/utils.rs:
```diff
@@ -353,11 +276,12 @@ pub(crate) async fn gen_wallets_transfer_tssc(
-    let mut wallet_addresses = Vec::<Address>::new();
-    let mut wallet_priv_keys = Vec::<String>::new();
-    let mut signers: Vec<Wallet<SigningKey>> = Vec::new();
+    let mut wallet_addresses = Vec::<Address>::with_capacity(num_accounts as usize);
```
Generally speaking I like to avoid mutation which I believe could be done here with a `map` function.

abhi3700 replied:
[commit](https://github.com/abhi3700/domain-transaction-producer/pull/5/commits/f0dc790970d039fca4c54e02af52285c1574ab3a)

0xjeremyfrank commented on dtp/src/main.rs:
```diff
@@ -150,10 +157,23 @@ async fn main() -> Result<()> {
+                        println!("Sending heavy transactions...");
+                        // Approach-2: All new wallet accounts are sender for each call individually
+                        // Say, all of them want to increment
+                        multicall_heavy_txs_2(
```
Is there an approach 1 we would use? Same question for light txs?

abhi3700 replied:
Yes there is one.
At this point, I am a little dubious about its reconsideration. It's mostly going to be entirely removed and the functions are going to be renamed accordingly. The focus here is to have more efficient concurrent calls (async or parallelism)....planned for PR #6. Until then in my opinion, let's keep as is. Although the approach-1 is archived for now.

0xjeremyfrank commented on dtp/src/contracts.rs:
```diff
@@ -0,0 +1,138 @@
+        "\'{}\' set Array with count {}, which incurred a gas fee of \'{} TSSC\', has a tx hash: \'{:?}\', indexed at #{} in block #{}.\n",
+        tx_receipt.from,
+        count,
+        get_gas_cost(&client, &tx_receipt).await?,
```
I think this call is wasteful and that we don't really need this  info.

abhi3700 replied:
This info is although showing as log when run in debug mode, otherwise not. Do you still want me to remove this `get_gas_cost` or `info!(...)` function call altogether.

0xjeremyfrank replied:
wouldn't `debug!` be what you are looking for here?

abhi3700 replied:
Initially, I was dubious about their usage initially. But when I learnt more about the differences in usage, the current details are not too verbose. So, used `info!` instead of `debug!`.

This is the info I found as their difference when to use:
<details>
<summary>See the differences:</summary>
Example:

Consider an application that processes user transactions:

- You might use `log::info!` to indicate that the application has started processing a batch of transactions.
- You would use `log::debug!` to log the details of each transaction in the batch, which might include intermediate states, specific values being processed, or decisions made by the transaction processing logic.

</details>

0xjeremyfrank replied:
My concern is the async calls that happen within `get_gas_cost`, not so much the verbosity.

abhi3700 replied:
[commit](https://github.com/subspace/domain-transaction-producer/pull/5/commits/f0eb6cd01dd1078bde88441cdbc6199e7fccb11d)

0xjeremyfrank reviewed (commented):
Things are improving. I still think the code has a lot of repetition and could be more efficient. I left a few comments that could be generalized to multiple places.

abhi3700 commented:
> Things are improving. I still think the code has a lot of repetition and could be more efficient.

Yeah I 💯 agree with you, it's just that my primary focus is to bring on all the features live and then focus on code refactoring ensuring efficiency in performance. But, please don't hesitate to give constant feedbacks, that I will definitely add alongside each PR, unless the commit is too long (pushed to next PR).
In parallel, I am constantly doing my own research to better the async calls (also getting to learn more on Async Rust) done in DTP CLI App.

abhi3700 committed:
Added chain id as single async call instead as env var

This change is done to ensure there is no mismatch when setting wrong chain id in the '.env' file during making calls to RPC.

abhi3700 committed:
Modified function for generating wallets without using mut type collections

abhi3700 committed:
Removed 'contracts' submodule

0xjeremyfrank commented on dtp/src/utils.rs:
```diff
@@ -272,44 +305,51 @@ pub(crate) async fn gen_wallets_transfer_tssc(
+
+    // TODO: [OPTIONAL] save the keypair into a local file or show in the output. Create a CLI flag like --to-console/--to-file
+    // Extract and format the private keys of the wallets for logging purposes
+    let wallet_priv_keys = wallets
```
you can extract the addresses and keys in one `map` function rather than iterating over `wallets` twice.

abhi3700 replied:
[commit](https://github.com/subspace/domain-transaction-producer/pull/5/commits/f2584681563d0857a63f8af22bf510bfdd56df2c).

I have something to put here as note:

Using std. rust, there are 2 possible approaches due to limitation of `unzip()` i.e. it supports a tuple of pairs, but not more than 2 elements in a tuple:
- A-1: Use (Vec<_>, Hashmap<address, priv_key>) as tuple of pairs for generating wallets. And then extract addresses from the hashmap keys to feed into `transfer_tssc_bulk` function.
- A-2: Use (Vec<_>, Vec<(address, priv_key)>) as tuple of pairs for generating wallets. And then extract addresses from the Vec keys to feed into `transfer_tssc_bulk` function.

-----
Now, using non-std. rust i.e. external crate `itertools`, `multiunzip()` method could be used to extract all 3 as separate vectors in 1 loop. This is the one implemented in this commit.

---
In summary, all these aforementioned approaches do use at least 2 loops.

Therefore, I think mutating all Vec and populating them via a single iteration as was done originally is more efficient (from looping standpoint at least) as there was just 1 looping done.

Would like to know your thoughts on this?

0xjeremyfrank replied:
why are we extracting `priv_keys` at all? I don't think they are used. I would probably just return `wallets` then do another map over wallets to pass addresses to `transfer_tssc_bulk`. Also, we may be doing too much in this function, why not have a `generate_wallets` function then outside of that function call `transfer_tssc_bulk`?

abhi3700 replied:
Yeah, we can move the private keys to some files, if needed.
Considering this, the new change has only 1 loop and extracting wallet & addresses. For now, just displaying the address & private keys for each wallets.

[commit](https://github.com/subspace/domain-transaction-producer/pull/5/commits/f0eb6cd01dd1078bde88441cdbc6199e7fccb11d)

abhi3700 committed:
Minimize iteration for wallet, addresses, private keys

Using std. rust, there are 2 possible approaches due to limitation of `unzip()` i.e. it supports a tuple of pairs, but not more than 2 elements in a tuple:
- A-1: Use (Vec<_>, Hashmap<address, priv_key>) as tuple of pairs for generating wallets. And then extract addresses from the hashmap keys to feed into `transfer_tssc_bulk` function.
- A-1: Use (Vec<_>, Vec<(address, priv_key)>) as tuple of pairs for generating wallets. And then extract addresses from the Vec keys to feed into `transfer_tssc_bulk` function.

-----
Now, using non-std. rust i.e. external crate `itertools`, `multiunzip()` method could be used to extract all 3 as separate vectors in 1 loop. This is the one implemented in this commit.

---
In summary, all these aforementioned approaches do use at least 2 loops.

Therefore, I think mutating all Vec and populating them via a single iteration as was done originally is better as there is just 1 looping done.

abhi3700 committed:
Code optimization

- Removed 'priv_keys' Vec collection during wallet, addresses generation. Later on, the private keys might be copied into some file.
- Removed 'itertools' external crate associated code.
- Removed 'get_gas_cost' (in TSSC) function async call in every light/heavy transactions sending processes. Not required as the purpose is to focus on load testing of EVM network. So, instead shown gas unit that each call incurs.
- Replaced 'info' w 'debug' logging when showing the tx details during sending light/heavy txs.

abhi3700 committed:
Added related code for easy debugging for heavy tx type

Some code refactored to shorten the code size a little bit. Logs for heavy txs added for gemini-3g testnet

0xjeremyfrank reviewed (approved):
I think the next PR should be to do some further cleanup of the code.

abhi3700 commented:
> I think the next PR should be to do some further cleanup of the code.

Yeah indeed.

Can I merge this PR?
